File: packages/depz-sensor-sdk-csharp/src/Depz.Sensor/Vl53l8/Advanced.cs
using System.Buffers.Binary;

namespace Depz.Sensor.Vl53l8;

/// <summary>
/// VL53L8 advanced-feature DCI codecs — pure, verifiable ports of the ST ULD
/// (BSD-3-Clause) plugin encoders: motion-indicator configuration,
/// detection-threshold block, and xtalk-margin scaling. Shared across both ToF
/// variants (VL53L8CX and VL53L8CH; see <see cref="Vl53l8Variant"/>). The live
/// DCI read/write transport that carries these to the sensor is
/// hardware-dependent and out of scope (see <see cref="Vl53l8Uld"/>).
/// </summary>
public static class Vl53l8Advanced
{
    public const int Resolution4x4 = 16;
    public const int Resolution8x8 = 64;

    // Detection-threshold measurement selectors + scale factors (plugin source):
    // get divides, set multiplies by these.
    public const int DistMm = 1;
    public const int SignalPerSpadKcps = 2;
    public const int RangeSigmaMm = 4;
    public const int AmbientPerSpadKcps = 8;
    public const int NbSpadsEnabled = 13;
    public const int MotionIndicator = 19;

    private static readonly IReadOnlyDictionary<int, int> ThreshScale = new Dictionary<int, int>
    {
        [DistMm] = 4,
        [SignalPerSpadKcps] = 2048,
        [RangeSigmaMm] = 128,
        [AmbientPerSpadKcps] = 2048,
        [NbSpadsEnabled] = 256,
        [MotionIndicator] = 65535,
    };

    public const int NbThresholds = 64;

    /// <summary>Xtalk margin raw DCI value: round(kcps × 2048).</summary>
    public static uint XtalkMarginRaw(double marginKcps) =>
        (uint)(long)Math.Round(marginKcps * 2048.0, MidpointRounding.AwayFromZero);

    /// <summary>One detection threshold (raw real-unit low/high, before scaling).</summary>
    public sealed record DetectionThreshold(
        int LowThresh, int HighThresh, int Measurement, int Type, int ZoneNum, int Operation);

    /// <summary>The 8-byte DCI_DET_THRESH_VALID_STATUS payload written before the block.</summary>
    public static byte[] ThresholdValidStatus() => new byte[] { 5, 5, 5, 5, 5, 5, 5, 5 };

    /// <summary>
    /// Encode the 64×12-byte DCI_DET_THRESH_START block. Each entry is packed as
    /// (i32 low, i32 high, u8 measurement, u8 type, u8 zone, u8 op) with low/high
    /// multiplied by the measurement's scale factor; missing entries are zeros.
    /// </summary>
    public static byte[] PackDetectionThresholds(IReadOnlyList<DetectionThreshold> thresholds)
    {
        var packed = new byte[NbThresholds * 12];
        for (int k = 0; k < NbThresholds; k++)
        {
            if (k >= thresholds.Count)
                continue;
            DetectionThreshold t = thresholds[k];
            int scale = ThreshScale.TryGetValue(t.Measurement, out int sc) ? sc : 1;
            int lo = t.LowThresh * scale;
            int hi = t.HighThresh * scale;
            int b = k * 12;
            BinaryPrimitives.WriteInt32LittleEndian(packed.AsSpan(b), lo);
            BinaryPrimitives.WriteInt32LittleEndian(packed.AsSpan(b + 4), hi);
            packed[b + 8] = (byte)t.Measurement;
            packed[b + 9] = (byte)t.Type;
            packed[b + 10] = (byte)t.ZoneNum;
            packed[b + 11] = (byte)t.Operation;
        }
        return packed;
    }
}

/// <summary>
/// Mirror of VL53L8CX_Motion_Configuration (156 bytes, ST ULD motion plugin).
/// <see cref="Pack"/> is byte-exact with the C struct
/// <c>&lt;i 3I 12B 64b 32B 32B&gt;</c>.
/// </summary>
public sealed class MotionConfig
{
    public int RefBinOffset;
    public uint DetectionThreshold;
    public uint ExtraNoiseSigma;
    public uint NullDenClipValue;
    public byte MemUpdateMode;
    public byte MemUpdateChoice;
    public byte SumSpan;
    public byte FeatureLength;
    public byte NbOfAggregates;
    public byte NbOfTemporalAccumulations;
    public byte MinNbForGlobalDetection;
    public byte GlobalIndicatorFormat1;
    public byte GlobalIndicatorFormat2;
    public byte Spare1;
    public byte Spare2;
    public byte Spare3;
    public readonly sbyte[] MapId = new sbyte[64];
    public readonly byte[] IndicatorFormat1 = new byte[32];
    public readonly byte[] IndicatorFormat2 = new byte[32];

    /// <summary>Serialize to the 156-byte VL53L8CX_Motion_Configuration wire form.</summary>
    public byte[] Pack()
    {
        var buf = new byte[156];
        BinaryPrimitives.WriteInt32LittleEndian(buf.AsSpan(0), RefBinOffset);
        BinaryPrimitives.WriteUInt32LittleEndian(buf.AsSpan(4), DetectionThreshold);
        BinaryPrimitives.WriteUInt32LittleEndian(buf.AsSpan(8), ExtraNoiseSigma);
        BinaryPrimitives.WriteUInt32LittleEndian(buf.AsSpan(12), NullDenClipValue);
        buf[16] = MemUpdateMode;
        buf[17] = MemUpdateChoice;
        buf[18] = SumSpan;
        buf[19] = FeatureLength;
        buf[20] = NbOfAggregates;
        buf[21] = NbOfTemporalAccumulations;
        buf[22] = MinNbForGlobalDetection;
        buf[23] = GlobalIndicatorFormat1;
        buf[24] = GlobalIndicatorFormat2;
        buf[25] = Spare1;
        buf[26] = Spare2;
        buf[27] = Spare3;
        for (int k = 0; k < 64; k++)
            buf[28 + k] = (byte)MapId[k];
        for (int k = 0; k < 32; k++)
            buf[92 + k] = IndicatorFormat1[k];
        for (int k = 0; k < 32; k++)
            buf[124 + k] = IndicatorFormat2[k];
        return buf;
    }

    /// <summary>
    /// The default motion configuration produced by vl53l8cx_motion_indicator_init
    /// for the given resolution (distance-window preset + zone map).
    /// </summary>
    public static MotionConfig InitDefault(int resolution)
    {
        var cfg = new MotionConfig
        {
            RefBinOffset = 13633,
            DetectionThreshold = 2883584,
            ExtraNoiseSigma = 0,
            NullDenClipValue = 0,
            MemUpdateMode = 6,
            MemUpdateChoice = 2,
            SumSpan = 4,
            FeatureLength = 9,
            NbOfAggregates = 16,
            NbOfTemporalAccumulations = 16,
            MinNbForGlobalDetection = 1,
            GlobalIndicatorFormat1 = 8,
            GlobalIndicatorFormat2 = 0,
        };
        cfg.SetResolution(resolution);
        return cfg;
    }

    /// <summary>vl53l8cx_motion_indicator_set_resolution: fill the per-zone map id.</summary>
    public void SetResolution(int resolution)
    {
        if (resolution == Vl53l8Advanced.Resolution4x4)
        {
            for (int i = 0; i < 16; i++)
                MapId[i] = (sbyte)i;
            for (int i = 16; i < 64; i++)
                MapId[i] = -1;
        }
        else if (resolution == Vl53l8Advanced.Resolution8x8)
        {
            for (int i = 0; i < 64; i++)
                MapId[i] = (sbyte)(((i % 8) / 2) + (4 * (i / 16)));
        }
        else
        {
            throw new ArgumentException($"resolution must be 16 or 64, got {resolution}", nameof(resolution));
        }
    }
}

File: packages/depz-sensor-sdk-rust/src/vl53l8/advanced.rs
//! VL53L8 advanced-feature DCI codecs (UM3109 plugins, contract 04).
//!
//! Shared by both ToF variants (VL53L8CX and VL53L8CH): the advanced DCI byte
//! layouts are identical, so there is one set of codecs, not a per-variant pair.
//!
//! Pure encoders ported verbatim from the ST ULD (BSD-3): the motion-indicator
//! configuration struct, the crosstalk-margin scaling, and the 64-entry
//! detection-threshold block. These are the byte layouts written over the DCI;
//! the live register bridge that performs the writes is out of scope.

use super::decode::{RESOLUTION_4X4, RESOLUTION_8X8};

/// Detection-threshold measurement selectors (`measurement` field).
pub const DIST_MM: u8 = 1;
pub const SIGNAL_PER_SPAD_KCPS: u8 = 2;
pub const RANGE_SIGMA_MM: u8 = 4;
pub const AMBIENT_PER_SPAD_KCPS: u8 = 8;
pub const NB_TARGET_DETECTED: u8 = 9;
pub const TAR_STATUS: u8 = 12;
pub const NB_SPADS_ENABLED: u8 = 13;
pub const MOTION_INDICATOR: u8 = 19;

pub const NB_THRESHOLDS: usize = 64;

/// Power modes (`vl53l8cx_api.h`).
pub const POWER_MODE_SLEEP: u8 = 0;
pub const POWER_MODE_WAKEUP: u8 = 1;
pub const POWER_MODE_DEEP_SLEEP: u8 = 2;

/// Per-measurement fixed-point scale applied to threshold low/high on set
/// (get divides by the same factor); unlisted measurements scale by 1.
fn thresh_scale(measurement: u8) -> i64 {
    match measurement {
        DIST_MM => 4,
        SIGNAL_PER_SPAD_KCPS => 2048,
        RANGE_SIGMA_MM => 128,
        AMBIENT_PER_SPAD_KCPS => 2048,
        NB_SPADS_ENABLED => 256,
        MOTION_INDICATOR => 65535,
        _ => 1,
    }
}

/// Xtalk margin (kcps/SPAD) → raw DCI value: `round(kcps * 2048)`.
pub fn xtalk_margin_to_raw(margin_kcps: f64) -> u32 {
    (margin_kcps * 2048.0).round() as u32
}

/// Mirror of `VL53L8CX_Motion_Configuration` (156 bytes, plugin source).
/// `pack()` reproduces the C struct byte layout (`<i3I12B64b32B32B`).
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct MotionConfig {
    pub ref_bin_offset: i32,
    pub detection_threshold: u32,
    pub extra_noise_sigma: u32,
    pub null_den_clip_value: u32,
    pub mem_update_mode: u8,
    pub mem_update_choice: u8,
    pub sum_span: u8,
    pub feature_length: u8,
    pub nb_of_aggregates: u8,
    pub nb_of_temporal_accumulations: u8,
    pub min_nb_for_global_detection: u8,
    pub global_indicator_format_1: u8,
    pub global_indicator_format_2: u8,
    pub spare1: u8,
    pub spare2: u8,
    pub spare3: u8,
    pub map_id: [i8; 64],
    pub indicator_format_1: [u8; 32],
    pub indicator_format_2: [u8; 32],
}

impl Default for MotionConfig {
    fn default() -> Self {
        MotionConfig {
            ref_bin_offset: 0,
            detection_threshold: 0,
            extra_noise_sigma: 0,
            null_den_clip_value: 0,
            mem_update_mode: 0,
            mem_update_choice: 0,
            sum_span: 0,
            feature_length: 0,
            nb_of_aggregates: 0,
            nb_of_temporal_accumulations: 0,
            min_nb_for_global_detection: 0,
            global_indicator_format_1: 0,
            global_indicator_format_2: 0,
            spare1: 0,
            spare2: 0,
            spare3: 0,
            map_id: [0; 64],
            indicator_format_1: [0; 32],
            indicator_format_2: [0; 32],
        }
    }
}

impl MotionConfig {
    /// Serialize to the 156-byte DCI payload.
    pub fn pack(&self) -> Vec<u8> {
        let mut out = Vec::with_capacity(156);
        out.extend_from_slice(&self.ref_bin_offset.to_le_bytes());
        out.extend_from_slice(&self.detection_threshold.to_le_bytes());
        out.extend_from_slice(&self.extra_noise_sigma.to_le_bytes());
        out.extend_from_slice(&self.null_den_clip_value.to_le_bytes());
        out.push(self.mem_update_mode);
        out.push(self.mem_update_choice);
        out.push(self.sum_span);
        out.push(self.feature_length);
        out.push(self.nb_of_aggregates);
        out.push(self.nb_of_temporal_accumulations);
        out.push(self.min_nb_for_global_detection);
        out.push(self.global_indicator_format_1);
        out.push(self.global_indicator_format_2);
        out.push(self.spare1);
        out.push(self.spare2);
        out.push(self.spare3);
        for b in &self.map_id {
            out.push(*b as u8);
        }
        out.extend_from_slice(&self.indicator_format_1);
        out.extend_from_slice(&self.indicator_format_2);
        out
    }

    /// Set `map_id` for the resolution (pure; mirrors `_set_resolution`).
    pub fn set_resolution(&mut self, resolution: usize) {
        if resolution == RESOLUTION_4X4 {
            for i in 0..16 {
                self.map_id[i] = i as i8;
            }
            for i in 16..64 {
                self.map_id[i] = -1;
            }
        } else if resolution == RESOLUTION_8X8 {
            for i in 0..64 {
                self.map_id[i] = (((i % 8) >> 1) + 4 * (i / 16)) as i8;
            }
        }
    }
}

/// The default motion-indicator configuration `motion_indicator_init` programs
/// for a resolution (the exact bytes written to the sensor).
pub fn default_motion_config(resolution: usize) -> MotionConfig {
    let mut cfg = MotionConfig {
        ref_bin_offset: 13633,
        detection_threshold: 2_883_584,
        extra_noise_sigma: 0,
        null_den_clip_value: 0,
        mem_update_mode: 6,
        mem_update_choice: 2,
        sum_span: 4,
        feature_length: 9,
        nb_of_aggregates: 16,
        nb_of_temporal_accumulations: 16,
        min_nb_for_global_detection: 1,
        global_indicator_format_1: 8,
        global_indicator_format_2: 0,
        ..MotionConfig::default()
    };
    cfg.set_resolution(resolution);
    cfg
}

/// One detection-threshold entry (real units; scaled on pack).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
pub struct DetectionThreshold {
    pub low_thresh: i32,
    pub high_thresh: i32,
    pub measurement: u8,
    pub type_: u8,
    pub zone_num: u8,
    pub operation: u8,
}

/// The two DCI blocks written by `set_detection_thresholds`.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct DetectionThresholdBlocks {
    /// `DCI_DET_THRESH_START` payload: 64 × 12 bytes.
    pub start: Vec<u8>,
    /// `DCI_DET_THRESH_VALID_STATUS`: 8 bytes, all `0x05`.
    pub valid: Vec<u8>,
}

/// Pack up to 64 detection thresholds into their DCI blocks. Entries beyond
/// the supplied list are zero-filled. Each entry's low/high are multiplied by
/// its measurement's fixed-point scale.
pub fn pack_detection_thresholds(thresholds: &[DetectionThreshold]) -> DetectionThresholdBlocks {
    let valid = vec![0x05u8; 8];
    let mut start = vec![0u8; NB_THRESHOLDS * 12];
    for k in 0..NB_THRESHOLDS {
        let t = thresholds.get(k).copied().unwrap_or_default();
        let scale = thresh_scale(t.measurement);
        let low = (t.low_thresh as i64 * scale) as i32;
        let high = (t.high_thresh as i64 * scale) as i32;
        let off = k * 12;
        start[off..off + 4].copy_from_slice(&low.to_le_bytes());
        start[off + 4..off + 8].copy_from_slice(&high.to_le_bytes());
        start[off + 8] = t.measurement;
        start[off + 9] = t.type_;
        start[off + 10] = t.zone_num;
        start[off + 11] = t.operation;
    }
    DetectionThresholdBlocks { start, valid }
}
